### PR TITLE
Lesson 4

### PR DESCRIPTION
**Pull request by Jarchik**
Users can select a language, course, get grade only once.
Login/Logout functionality
DB integration
Templates

### CODE REVIEW

File: lesson_4/main.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+class Course(): 
+    def __init__(self, course_id: str, name: str) -> None:
+        self.id = course_id
+        self.name = name
+
+
+class UserModel():
+    def create_user(self, user_name: str, user_lang: str) -> Union[bool, ProgrammingCoursesException]:
+        try:
+            sql = f"INSERT INTO user(Name, Language, Course, Grade) VALUES('{user_name}', '{user_lang}', Null, Null);"
```
Comment (idegtiarov) [inline lesson_4/main.py:66]: U have lots of code duplication in every UserModel class method I would consider splitting it into an additional function or decorator:
to provide communication with the database.

File: lesson_4/main.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        self.courses = {}
+        self.student = None
+        self.user_model = user_model
+    
+    def add_course(self, course: Course):
+        self.courses[course.id] = course
+
+    def get_course_id_by_name(self, course_name: str) -> Union[str, ProgrammingCoursesException]:
+        correct_id = ''
+        for course_id, course in self.courses.items():
```
Comment (idegtiarov) [inline lesson_4/main.py:136]: It is not clear why the dict self.courses contain as keys, not a course_name, could you please explain your choice? Within course_name as a key the search complexity would be O(1) in your variant it is O(n)

**Review by idegtiarov — COMMENTED** [review]
Two comments in addition to the inline ones, 

- It would be great to see the documentation in the code with the description of the functions, classes, methods
- And a big request in the cases some code was already reviewed prepare PR with the diff not all files especially when we have a lot of them

All other looks great, thanks!

File: lesson_4/main.py (hunk 1 above)
Comment (Jarchik) [reply lesson_4/main.py:66]: Yeah, I thought about it too. But this code already took a lot of my time. Probably, will improve the logic in Django. :)

File: lesson_4/main.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    def menu_wrapper(*args, **kwargs):
+            return func(*args, **kwargs) + bottom
+    return menu_wrapper
+
+
+class ProgrammingCoursesException(Exception):
+    def __init__(self, msg: str):
+        self.error_msg = msg
+    def __str__(self):
+        return repr(self.error_msg)
```
[the remark this answers is not in the document]
Comment (Jarchik) [reply lesson_4/main.py:54]: Oh. It should be self.msg, I guess...

[the remark this answers is not in the document]
Comment (idegtiarov) [reply lesson_4/main.py:54]: not sure

**Pull request merged**